state-chain/pallets/cf-trading-strategy/src/lib.rs

```diff
@@ -0,0 +1,351 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub mod migrations;
+
+pub mod weights;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+
+use frame_support::sp_runtime::FixedU64;
+
+use cf_primitives::{Asset, AssetAmount, Tick, STABLE_ASSET};
+use cf_traits::{
+	AccountRoleRegistry, BalanceApi, Chainflip, IncreaseOrDecrease, NonceProvider, OrderId, Side,
+	TradingStrategyParameters, TradingStrategyParametersProvider,
+};
+
+pub use pallet::*;
+use weights::WeightInfo;
+
+pub const PALLET_VERSION: StorageVersion = StorageVersion::new(1);
+
+// Note that strategies can only create one order per asset/side so we can just
+// have a fixed order id (at least until we develop more advanced strategies).
+const STRATEGY_ORDER_ID: OrderId = 0;
+
+#[derive(Clone, Debug, Encode, Decode, TypeInfo, PartialEq, Eq)]
+struct TradingStrategyEntry<AccountId> {
+	base_asset: Asset,
+	owner: AccountId,
+	strategy: TradingStrategy,
+}
+
+#[derive(Clone, Debug, Encode, Decode, TypeInfo, PartialEq, Eq)]
+pub enum TradingStrategy {
+	SellAndBuyAtTicks { sell_tick: Tick, buy_tick: Tick },
+}
+
+fn derive_strategy_id<T: Config>(lp: &T::AccountId, nonce: T::Nonce) -> T::AccountId {
+	use frame_support::{sp_runtime::traits::TrailingZeroInput, Hashable};
+
+	// Combination of lp + nonce is unique for every successful call, so this should
+	// generate unique ids:
+	Decode::decode(&mut TrailingZeroInput::new(
+		(*b"chainflip/strategy_account", lp.clone(), nonce).blake2_256().as_ref(),
+	))
+	.unwrap()
```

Comment:
Why do we need trailing zeros here here? blake hashing will produce 32 bytes. Is it to protect against larger than 32 byte ids? This function doesn't protect against smaller than 32 byte ids - might be easier to just return an error if the blake hash doesn't decode into account id, if we want to cover all cases.

Reply:
I just copied this from other places where we did something similar (this should probably be a helper function). Yes, I think the only difference is that this won't fail if we used AccountId that's larger than 32 bytes. I would rather we didn't return Error and have callers deal with something that we should be able to handle internally. I think the most sane approach is to statically(?) assert that our id isn't larger than 32 bytes and remove `TrailingZeroInput`.

Reply:
I think that's also reasonable